haigha/transports/socket_transport.py

```diff
@@ -29,23 +29,33 @@ def connect(self, (host, port), klass=socket.socket):
         Connect assuming a host and port tuple.
         '''
         self._host = "%s:%s" % (host, port)
-        self._sock = klass()
+        infos = socket.getaddrinfo(host, port, 0, 0, socket.IPPROTO_TCP)
+        if not infos:
+            raise socket.error("getaddrinfo returns an empty list")
+
+        # Sort addresses such that the new IPv6 address is first.
+        for info in reversed(sorted(infos, key=lambda info: len(info[-1]))):
+
+            sockaddr = info[-1]
+            try:
+                # Check if the IP is a valid IPv6 address. Fallback to IPv4 if not.
+                socket.inet_pton(socket.AF_INET6, sockaddr[0])
+                self._sock = klass(socket.AF_INET6)
+                break
+            except socket.error:
+                self._sock = klass()
+                break
+        else:
+
+            raise socket.error("No valid socket object could be created.")
```

Comment:
As above, let's not raise the socket module's exception. We should use haigha's own exception. The problem here is that the higher level module (and the engineer who is debugging) doesn't get an opportunity to find out the underlying error and error message that may be critical for resolving the issue.

A good approach may be to track how many have been tried, and if we find ourselves in the `except socket.error` block having just tried the last address, simply re-raise that last exception directly, preserving the traceback, socket.error.errno, and error message.

Logging the excpetion values would be extremely helpful here, too: perhaps at debug level for intermediate failures and at error level for the final failure (when there are no more addresses to try)